### PR TITLE
[Events] Basic cluster events for autoscaler

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This adds the basic cluster events to the autoscaler. In order to properly support this feature, we need some additional work, but I think it is a good first iterative steps. In the future, we need

1. Add severity to message.
2. Refactoring event summarize and global event system
3. Support events emitted from CLI (e.g., global_event_system).
cc @wuisawesome 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by wuisawesome — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Comment by wuisawesome** [comment]
Uh oh operator autoscaling test is failing, i think it's sometimes flakey, but also very related. let me try rerunning it, but i'm scared there's a genuine issue

**Comment by wuisawesome** [comment]
@DmitriGekhtman wdyt?

**Comment by DmitriGekhtman** [comment]
@DmitriGekhtman thinks indeed we should double-check the operator test -- it makes some assertions about autoscaler events being output to logs.

@DmitriGekhtman is also pretty tired and checked out for the week and would like to follow up with a thoughtful review on Tuesday.

Are we trying to get this in for Ray 2.1.0?

**Comment by rkooo567** [comment]
yeah I hoped, but I think some test failures are still related (btw this PR wouldn't impact anything to the existing log output), so I think I cannot make it by 2.1

File: python/ray/tests/test_state_api.py [hunk 1]
```diff
@@ -1851,33 +1852,32 @@ def verify():
...
+        ) in events[0]["message"]
+        return True
 
-#     wait_for_condition(verify)
-#     print(list_cluster_events())
+    wait_for_condition(verify)
+    print(list_cluster_events())
 
-#     # TODO(sang): Support get_cluster_events
+    # TODO(sang): Support get_cluster_events
```
Comment (DmitriGekhtman) [inline python/ray/tests/test_state_api.py:1880]: Did you want to remove this comment?
Or do we still need a get_cluster_event?

**Review by DmitriGekhtman — APPROVED** [review]
Looks good, question about comment.

Needs rebasing. I can help look into K8s test failures.

**Comment by DmitriGekhtman** [comment]
Looks like I wasn't paying attention during a design review -- will leave a futile complaint about continuing using logs for inter-process communication: 
https://github.com/ray-project/ray/issues/26186

This is fine, though.

**Comment by wuisawesome** [comment]
> Looks like I wasn't paying attention during a design review -- will leave a futile complaint about continuing using logs for inter-process communication: https://github.com/ray-project/ray/issues/26186

+1 to this

**Comment by rkooo567** [comment]
@DmitriGekhtman yeah it'd be great if you can help debugging test failures. 

Re; using logs to publish events. It should be easy to switch the implementation. I think for other components, there's not many caveat using log files, but I agree it's makes sketchy things when doing the same thing from the autoscaler (because this reading logs -> report has assumption components are in the same container, but autocaler doesn't follow that assumption). The interface is written in a way it's easy to change later (and I think @wuisawesome will take a look). We should just make sure we don't report individual event log as RPC so it will overload dashboard (we need some sort of buffering)

**Comment by DmitriGekhtman** [comment]
After it's rebased, I will take a look at test failures.

**Comment by rkooo567** [comment]
@DmitriGekhtman rebased is done. I will take a look at other test failure soon.

File: python/ray/tests/test_state_api.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/tests/test_state_api.py:1880]: we still need a get cluster event

**Comment by rkooo567** [comment]
@DmitriGekhtman do you know why the operator test fails?

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman do you know why the operator test fails?

It appears to be timing out. I will investigate.

**Comment by DmitriGekhtman** [comment]
I attempted to reproduce the issue by using an image with the code changes, but I was not successful.
I could try to reproduce the problem in the CI environment, but could you rebase just one more time to make sure it's not some sort of random flakiness?

**Comment by rkooo567** [comment]
I am sure it is not random flakiness. Because I merged PRs many times, and this always failed. But I will merge the latest master one more time

**Comment by DmitriGekhtman** [comment]
> I am sure it is not random flakiness. Because I merged PRs many times, and this always failed. But I will merge the latest master one more time 

Got it, thanks for data points. Will set up the CI repro environment to figure out what's up.

**Comment by DmitriGekhtman** [comment]
Found the issue -- the autoscaler container is crashlooping with
```
Traceback (most recent call last):
  File "/home/ray/anaconda3/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/scripts/scripts.py", line 2596, in main
    return cli()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ray/anaconda3/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/scripts/scripts.py", line 2340, in kuberay_autoscaler
    from ray.autoscaler._private.kuberay.run_autoscaler import run_kuberay_autoscaler
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/kuberay/run_autoscaler.py", line 12, in <module>
    from ray.autoscaler._private.monitor import Monitor
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 18, in <module>
    from ray._private.event.event_logger import get_event_logger
ModuleNotFoundError: No module named 'ray._private.event'
```

Let me look into a fix, and also come up with a better way to surface these things.

**Comment by DmitriGekhtman** [comment]
Perhaps we need an __init__.py

**Comment by DmitriGekhtman** [comment]
@rkooo567 just pushed the init.py to your branch, let's see if that fixes it or not

**Comment by DmitriGekhtman** [comment]
> @rkooo567 just pushed the init.py to your branch, let's see if that fixes it or not

It appears to have done the trick.
I will merge by EOD, assuming no other CI issues.

**Comment by DmitriGekhtman** [comment]
The cause of the hang is fixed here:
https://github.com/ray-project/ray/pull/29907/files

**Comment by DmitriGekhtman** [comment]
Test failures are unrelated. Merging.

**Comment by rkooo567** [comment]
Awesome Thank you so much for handling this :)!! @DmitriGekhtman !!

**Pull request merged**